Code/Editor/EditorFramework/Assets/Implementation/AssetCurator.cpp

```diff
@@ -1553,39 +1496,112 @@ ezSubAsset* ezAssetCurator::GetSubAssetInternal(const ezUuid& assetGuid)
+void ezAssetCurator::BuildFileExtensionSet(ezSet<ezString>& AllExtensions)
 {
-  CURATOR_PROFILE("HandleSingleFile");
-  EZ_LOCK(m_CuratorMutex);
+  ezStringBuilder sTemp;
+  AllExtensions.Clear();
+
+  const auto& assetTypes = ezAssetDocumentManager::GetAllDocumentDescriptors();
 
-  ezFileStats Stats;
-  if (ezOSFile::GetFileStats(sAbsolutePath, Stats).Failed())
+  // use translated strings
+  ezMap<ezString, const ezDocumentTypeDescriptor*> allDesc;
+  for (auto it : assetTypes)
   {
-    // this is a bit tricky:
-    // when the document is deleted on disk, it would be nicer not to close it (discarding modifications!)
-    // instead we could set it as modified
-    // but then when it was only moved or renamed that means we have another document with the same GUID
-    // so once the user would save the now modified document, we would end up with two documents with the same GUID
-    // so, for now, since this is probably a rare case anyway, we just close the document without asking
-    ezDocumentManager::EnsureDocumentIsClosedInAllManagers(sAbsolutePath);
+    allDesc[ezTranslate(it.Key())] = it.Value();
+  }
 
-    if (ezFileStatus* pFileStatus = m_ReferencedFiles.GetValue(sAbsolutePath))
+  for (auto it : allDesc)
+  {
+    const auto desc = it.Value();
+
+    if (desc->m_pManager->GetDynamicRTTI()->IsDerivedFrom<ezAssetDocumentManager>())
     {
-      pFileStatus->m_Timestamp.Invalidate();
-      pFileStatus->m_uiHash = 0;
-      pFileStatus->m_Status = ezFileStatus::Status::Unknown;
+      sTemp = desc->m_sFileExtension;
+      sTemp.ToLower();
 
-      ezUuid guid0 = pFileStatus->m_AssetGuid;
-      if (guid0.IsValid())
+      AllExtensions.Insert(sTemp);
+    }
+  }
+}
+
+void ezAssetCurator::OnAssetFilesEvent(const ezFileChangedEvent& e)
+{
+  switch (e.m_Type)
+  {
+    case ezFileChangedEvent::Type::DocumentLinked:
+    {
+    }
+    break;
+    case ezFileChangedEvent::Type::FileAdded:
+    case ezFileChangedEvent::Type::FileChanged:
+    {
+      // If the asset was just added it is not tracked and thus no need to invalidate anything.
+      if (e.m_Type == ezFileChangedEvent::Type::FileChanged)
+      {
+        ezUuid guid0 = e.m_Status.m_DocumentID;
+        if (guid0.IsValid())
+          InvalidateAssetTransformState(guid0);
+
+        auto it = m_InverseTransformDeps.Find(e.m_sPath);
+        if (it.IsValid())
+        {
+          for (const ezUuid& guid : it.Value())
+          {
+            InvalidateAssetTransformState(guid);
+          }
+        }
+
+        auto it2 = m_InverseThumbnailDeps.Find(e.m_sPath);
+        if (it2.IsValid())
+        {
+          for (const ezUuid& guid : it2.Value())
+          {
+            InvalidateAssetTransformState(guid);
+          }
+        }
+      }
+
+      // Assets should never be in an AssetCache folder.
+      const char* szNeedle = e.m_sPath.FindSubString("AssetCache/");
+      if (szNeedle != nullptr && e.m_sPath.GetData() != szNeedle && szNeedle[-1] == '/')
```

Comment:
I can't remember why I implemented it like this. Now the input is always an absolute path, so if AssetCache is in the path, it will be in the form of "/AssetCache/" so I just search for that now.